src/core/Form/FileInput/FileInput.tsx

```diff
@@ -0,0 +1,603 @@
+    status,
+    statusText,
+    hintText,
+    id,
+    fullWidth,
+    forwardedRef,
+    statusTextAriaLiveMode = 'assertive',
+    'aria-describedby': ariaDescribedBy,
+    tooltipComponent,
+    globalMargins,
+    ...rest
+  } = props;
+  const [_marginProps, passProps] = separateMarginProps(rest);
+
+  const dragAreaRef = useRef<HTMLDivElement>(null);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const buildFileItemRefs = (fileList: FileList) => {
+    const newFileItemRefs = [];
+    for (let index = 0; index < fileList.length; index += 1) {
+      const file = fileList[index];
+      const fnRef = filePreview
+        ? createRef<HTMLAnchorElement>()
+        : createRef<HTMLDivElement>();
+      newFileItemRefs.push({
+        id: `${id}-${file.name.replace(/\s/g, '')}`,
+        fileNameRef: fnRef,
+        fileSizeElementId: `${id}-${file.name.replace(/\s/g, '')}-fileSize`,
+      });
+    }
+    return newFileItemRefs;
+  };
+
+  const buildFileListFromControlledValueObjects = (
+    controlledValueObjects: ControlledFileItem[],
+  ) => {
+    const newFileList = new DataTransfer();
+    controlledValueObjects.forEach((fileItem) => {
+      newFileList.items.add(fileItem.file);
+    });
+    return newFileList.files;
+  };
+
+  const [files, setFiles] = useState<FileList | null>(
+    controlledValue
+      ? buildFileListFromControlledValueObjects(controlledValue)
+      : null,
+  );
+  const [fileItemRefs, setFileItemRefs] = useState<FileItemRefs[]>(
+    controlledValue
+      ? buildFileItemRefs(
+          buildFileListFromControlledValueObjects(controlledValue),
+        )
+      : [],
+  );
+  const [mockInputWrapperFocus, setMockInputWrapperFocus] = useState(false);
+
+  const labelId = `${id}-label`;
+  const hintTextId = `${id}-hintText`;
+  const statusTextId = `${id}-statusText`;
+  const multiFileListHeadingId = `${id}-multiFileListHeading`;
+
+  // Remove the possibility to have undefined forwardedRef as a parameter for forkRefs
+  const definedRef = forwardedRef || null;
+
+  const dragEnterEventHandler = (e: DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragAreaRef.current?.classList.add('highlight');
+  };
+  const dragOverEventHandler = (e: DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragAreaRef.current?.classList.add('highlight');
+  };
+  const dragLeaveEventHandler = (e: DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragAreaRef.current?.classList.remove('highlight');
+  };
+  const dropEventHandler = (e: DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragAreaRef.current?.classList.remove('highlight');
+    const dt = e.dataTransfer;
+    const filesFromInput = dt?.files;
+    if (!controlledValue && filesFromInput && inputRef.current?.files) {
+      addNewFiles(filesFromInput, true);
+    }
+    if (propOnChange) {
+      propOnChange(filesFromInput || new FileList());
+    }
+  };
```

Comment:
This could probably be refactored to avoid duplicate code. Like a general drag handler that sets or unsets the highlight class which could then be called for each scenario.

Reply:
Refactored this a bit as you suggested